### PR TITLE
Minor improvements

### PR DESCRIPTION
**Pull request by sukolenvo**
Remove deprecated api usage,
Windows custom frame decoration support

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+plugins {
```
Comment (sukolenvo) [inline build.gradle.kts:1]: Converted to kotlin script

File: build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  pluginName.set("branch-window-title")
+  // // Disables updating since-build attribute in plugin.xml
+  updateSinceUntilBuild.set(false)
+  downloadSources.set(true)
+  plugins.addAll("git4idea", "svn4idea")
+}
+
+tasks.withType(JavaCompile::class.java) {
+  options.isDeprecation = true
+  options.encoding = "UTF-8"
```
Comment (sukolenvo) [inline build.gradle.kts:40]: Define encoding to fix build on Win

File: build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+plugins {
+  java
+  idea
+  id("org.jetbrains.intellij") version "1.3.0"
+}
+
+// TODO (TH) should be publisher company
+group = "org.jetbrains"
+
+version = "0.6.2"
```
Comment (sukolenvo) [inline build.gradle.kts:10]: Incremented

File: build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+dependencies {
+  testImplementation("junit:junit:4.12")
+  testImplementation("org.mockito:mockito-core:2.21.0")
+  testImplementation("org.assertj:assertj-core:3.21.0")
+}
+
+intellij {
+  // IntelliJ IDEA releases: https://www.jetbrains.com/intellij-repository/releases e.g. IC-2019.3
+  version.set("2021.2.3")
```
Comment (sukolenvo) [inline build.gradle.kts:29]: Set to latest

File: src/main/java/io/wisetime/idea/branch/BranchHelper.java [hunk 5]
```diff
@@ -24,12 +26,19 @@ public String getCurrentBranchName() {
 
   public void onBranchChanged(String branchName) {
     currentBranchName = branchName;
-    updateFrameTitle();
+    ApplicationManager.getApplication().invokeLater(this::updateFrameTitle);
```
Comment (sukolenvo) [inline src/main/java/io/wisetime/idea/branch/BranchHelper.java:29]: Frame/project name changes should be done on UI thread. Sometimes git plugin sends branch change notifications from worker thread.

File: src/main/java/io/wisetime/idea/branch/BranchHelper.java [hunk 6]
```diff
@@ -24,12 +26,19 @@ public String getCurrentBranchName() {
...
   }
 
   private void updateFrameTitle() {
-    final String projectTitle = ServiceManager.getService(FrameTitleBuilder.class)
-        .getProjectTitle(project);
+    final String projectTitle = FrameTitleBuilder.getInstance().getProjectTitle(project);
+    SettingsState settings = SettingsState.getInstance();
+    if (settings.appendToProjectName) {
+      updateProjectTitle(projectTitle);
```
Comment (sukolenvo) [inline src/main/java/io/wisetime/idea/branch/BranchHelper.java:36]: We cannot rollback project name reliably - therefore if settings disabled, leave as is

File: src/main/java/io/wisetime/idea/branch/BranchHelper.java [hunk 7]
```diff
@@ -38,4 +47,12 @@ private void updateFrameTitle() {
     }
   }
 
+  private void updateProjectTitle(String projectTitle) {
+    if (project instanceof ProjectEx) {
+      ((ProjectEx) project).setProjectName(projectTitle);
```
Comment (sukolenvo) [inline src/main/java/io/wisetime/idea/branch/BranchHelper.java:52]: This updated window title on windows:
image
And few other places on all OS:
image
I've been using it on my machine for few days, branch name in recent list is helpful

File: src/main/java/io/wisetime/idea/branch/settings/SettingsPanel.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+public class SettingsPanel {
+
+  private final JBCheckBox windowTitle;
+  private final JBCheckBox projectName;
+  private JPanel root;
+
+  public SettingsPanel() {
+    root = new JPanel(new VerticalFlowLayout());
+    windowTitle = new JBCheckBox("Append VCS branch name to window title");
+    projectName = new JBCheckBox("Append VCS branch name to project name (Experimental)");
```
Comment (sukolenvo) [inline src/main/java/io/wisetime/idea/branch/settings/SettingsPanel.java:23]: image

File: src/main/java/io/wisetime/idea/branch/settings/SettingsState.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+ * @author vadym
+ */
+@State(
+    name = "io.wisetime.idea.branch.settings.SettingsState",
+    storages = @Storage("WTBranchWindowTitleSettings.xml")
+)
+public class SettingsState implements PersistentStateComponent<SettingsState> {
+
+  public boolean appendToWindowTitle = true;
+  public boolean appendToProjectName = IdeFrameDecorator.isCustomDecorationAvailable();
```
Comment (sukolenvo) [inline src/main/java/io/wisetime/idea/branch/settings/SettingsState.java:26]: By default project name is changed only on Windows, other OS will have previous behaviour - only win title changes

Comment (sukolenvo) [reply src/main/java/io/wisetime/idea/branch/settings/SettingsState.java:26]: Or to play it safe we could disable it by default for all OS and maybe update screenshots?

File: src/main/java/io/wisetime/idea/branch/BranchHelper.java (hunk 5 above)
Comment (sukolenvo) [reply src/main/java/io/wisetime/idea/branch/BranchHelper.java:29]: This should fix #5

Comment (tom-haines) [reply src/main/java/io/wisetime/idea/branch/BranchHelper.java:29]: this non-blocking seems cleaner

File: src/main/java/io/wisetime/idea/branch/BranchHelper.java (hunk 7 above)
Comment (tom-haines) [reply src/main/java/io/wisetime/idea/branch/BranchHelper.java:52]: nice!

File: src/main/java/io/wisetime/idea/branch/settings/SettingsPanel.java (hunk 8 above)
Comment (tom-haines) [reply src/main/java/io/wisetime/idea/branch/settings/SettingsPanel.java:23]: Should this be the radio button component instead?
https://jetbrains.design/intellij/controls/radio_button/

File: src/main/java/io/wisetime/idea/branch/settings/SettingsState.java (hunk 9 above)
Comment (tom-haines) [reply src/main/java/io/wisetime/idea/branch/settings/SettingsState.java:26]: Is it a valid state that both appendToWindowTitle and appendToProjectName are true?

File: src/main/java/io/wisetime/idea/branch/BranchHelper.java (hunk 6 above)
Comment (tom-haines) [reply src/main/java/io/wisetime/idea/branch/BranchHelper.java:36]: SGTM

**Review by tom-haines — CHANGES_REQUESTED** [review]
Great work!
Two pending queries to resolve as to what is considered a valid data state for the preference values.

File: src/main/java/io/wisetime/idea/branch/settings/SettingsPanel.java (hunk 8 above)
Comment (sukolenvo) [reply src/main/java/io/wisetime/idea/branch/settings/SettingsPanel.java:23]: We change 
1. just window title
or
2. window title and project name

If only project name is changed - window title will be not contain branch name

File: src/main/java/io/wisetime/idea/branch/settings/SettingsState.java (hunk 9 above)
Comment (sukolenvo) [reply src/main/java/io/wisetime/idea/branch/settings/SettingsState.java:26]: Yes, we change both. On windows user don't see window title. On mac/linux/apps that read window title, can see custom frame decoration view.

File: src/main/java/io/wisetime/idea/branch/settings/SettingsPanel.java (hunk 8 above)
Comment (tom-haines) [reply src/main/java/io/wisetime/idea/branch/settings/SettingsPanel.java:23]: SGTM 👍

**Review by tom-haines — APPROVED** [review]

**Pull request merged**